### PR TITLE
Add Stream-Style Serialization for XML

### PR DESCRIPTION
**Pull request by alzimmermsft**
Adds stream-style serialization code generation for XML-based models. Generating with stream-style serialization removes the use of Jackson Databind and Jackson Dataformat XML in models.

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java [hunk 1]
```diff
@@ -101,71 +121,165 @@ protected void addModelConstructorParameter(ClientModelProperty property,
...
 
-        writeToJson(classBlock, propertiesManager);
-        writeFromJson(classBlock, model, propertiesManager, settings);
+        if (settings.isGenerateXmlSerialization() && model.getXmlName() != null) {
+            writeToXml(classBlock, propertiesManager);
+            writeFromXml(classBlock, model, propertiesManager, settings);
+        } else {
+            writeToJson(classBlock, propertiesManager);
+            writeFromJson(classBlock, model, propertiesManager, settings);
+        }
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java:224]: nit, given the size of this file, wondering that if JSON and XML is orthogonal, will it worthwhile to put them to separate classes.

PS: refactor can happen anytime. No need to be coupled.

**Review by weidongxu-microsoft — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java [hunk 2]
```diff
@@ -426,7 +507,7 @@ private static void writeSuperTypeFromJson(JavaClass classBlock, ClientModel mod
                 "reader.nextToken();",
                 "if (\"" + discriminatorProperty.getSerializedName() + "\".equals(reader.getFieldName())) {",
                 "    reader.nextToken();",
-                "    discriminatorValue = reader.getStringValue();",
+                "    discriminatorValue = reader.getString();",
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java:510]: This was a change to `azure-json` given it's still in beta

File: javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java [hunk 3]
```diff
@@ -865,8 +873,8 @@ private static void generateHeaderCollectionDeserialization(List<ClientModelProp
 
     private static void generateHeaderDeserializationFunction(ClientModelProperty property, JavaBlock javaBlock) {
         IType wireType = property.getWireType();
-        boolean needsNullGuarding = (wireType instanceof ClassType && wireType != ClassType.String)
-            || wireType instanceof ArrayType;
+        boolean needsNullGuarding = wireType != ClassType.String &&
+            (wireType instanceof ArrayType || wireType instanceof ClassType || wireType instanceof EnumType || wireType instanceof GenericType);
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java:877]: `MapType` should be handled by `GenericType`

File: javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java (hunk 1 above)
Comment (alzimmermsft) [reply javagen/src/main/java/com/azure/autorest/template/StreamSerializationModelTemplate.java:224]: Yeah, I do want to begin having the same API with slightly different implementation for generating JSON vs XML. And agreed that it'll be a future TODO

**Pull request merged**